_posts/2020-10-16-kaitai-struct-v0.9-released.md

```diff
@@ -63,7 +65,7 @@ fixes a lot of bugs and includes quite a few infrastructure improvements.
   * C++: add C++11 mode
     * Add `--cpp-standard` CLI option: pass `--cpp-standard 11` to enable C++11 mode (`98` is default)
     * C++11 target:
-      * uses `#pragma once` (instead of `#ifndef FOO_H_` header guards)
+      * uses `#pragma once` (instead of `#ifndef FOO_H_` header guards) ([25fb1eee61d07bdc8b199445776836ce9a6606ef](https://github.com/kaitai-io/kaitai_struct_compiler/commit/25fb1eee61d07bdc8b199445776836ce9a6606ef))
```

Comment:
I'm pretty sure that 40 characters of random rubbish will only distract and disturb the user, and most likely noone will actually ever open that commit - it's just absolutely pointless for the expected audience.

To be frank, I was already quite sceptical about the **issue** references, because the ubiquitous _(#1234)_ brackets have also potentially a disturbing effect. The only reason why I left them somewhere is that there is usually some explanation of how to use the particular feature, so they serve as a temporary replacement for a proper documentation.

But you can see that whenever there _is_ the documentation available (e.g. the [little-endian bit ints](https://doc.kaitai.io/user_guide.html#bit-ints-le)), I use **([docs](https://doc.kaitai.io/user_guide.html#bit-ints-le))** instead of ~~([#155](https://github.com/kaitai-io/kaitai_struct/issues/155))~~.

Reply:
Some of these changes were a big surprise for me. I mean I am subscribed to the bug trackers, and I was really surprised to find these changes, since I haven't remembered them. So I searched the bug tracker, and found nothing. So, the links to comments convey the information:

1. The addition of the feature was not visible in the issue tracker
2. when, where and by whom it was introduced.
3. what exactly effect it does and how to use it.

Reply:
> So, the links to commits convey the following information sparing the users from digging the history themselves

Why would normal _users_ dig the history in the first place? What would be the reason?

> 1. The addition of the feature was not visible in the issue tracker

IMHO it makes sense to open an issue only if there's an intent to give people know that the particular bug fix or feature is being implemented, or if there is anything to discuss. If it's something small and obvious, creating an issue is redundant, and it would often take longer than the code implementation itself.

> 2. when, where and by whom it was introduced.

This is exactly what commands [`git annotate`](https://git-scm.com/docs/git-annotate) and [`git blame`](https://git-scm.com/docs/git-blame) are for. Finding the commit where `#pragma once` was introduced is actually quite simple:

1. Search GitHub for [`org:kaitai-io pragma once`](https://github.com/search?q=org%3Akaitai-io+pragma+once&type=code) and click the line number containing `usePragmaOnce` in the compiler source code (line 17 on the screenshot):

   ![image](https://user-images.githubusercontent.com/47499687/97483321-9229e380-1957-11eb-9de3-79c95594f00f.png)

2. Click the meatballs menu (3 horizontal dots) to the left from the line number and choose _View git blame_:

   ![image](https://user-images.githubusercontent.com/47499687/97484114-9e627080-1958-11eb-9a35-9417977b9245.png)

3. Click the commit link on the left to go to the the commit detail page: 

   ![image](https://user-images.githubusercontent.com/47499687/97484508-116be700-1959-11eb-8cf6-a32887d2dd7a.png)

> 3. what exactly effect it does and how to use it.

Yeah, this is a point we can agree on, that's definitely worth knowing for the user. This is what I meant with referencing issues as the _temporary replacement for documentation_ in https://github.com/kaitai-io/kaitai-io.github.io/pull/17#discussion_r507208279. But l don't get how linking to a commit with the source code helps here: do you really think that a typical user will go through the source code to understand how he's supposed to use the feature? I don't think so.

Moreover, just in context of the `#pragma once` , this argument doesn't make much sense - it's too simple and obvious to deserve any detailed docs. And one doesn't _use_ it directly, it's more of a generated code improvement, the user actually doesn't need to know about it. And if one doesn't understand what `pragma once` means, all they need to do is to [search it with Google](https://www.google.com/search?q=pragma+once).

Given it's _that_ much obvious and simple, it also undermines the need of knowing who, when and where introduced it.

Reply:
>Finding the commit where #pragma once was introduced is actually even easier using the GitHub web GUI:

This was what I did (and for other features too) :) But even if we use GH for it, it is digging git history, just GH-assisted.

>Why would normal users dig the history in the first place? What would be the reason?

Just to know the feature enough. Not only the fact that it exists, but also what exactly it does and what is it intended to be used for.

>do you really think that a typical user will go through the source code to understand how he's supposed to use the feature?

Yes. Digging into source code is not the best way, but if it is the only way (since there is no docs for some features), one has to. If one is determined to get known how to use the feature, lack of docs won't stop him.

>Moreover, just in context of the #pragma once , this argument doesn't make much sense - it's too simple and obvious to deserve any detailed docs.

It doesn't deserve the detailed docs, but a user seeing it deserves the clarification for the question "WTF? I am monitoring the repo, but I don't remember this feature. Have I messed anything?" So, a link to a commit is an answer "No, you haven't missed anything, it was just silently added to the repo".

Reply:
> "No, you haven't missed anything, it was just silently added to the repo".

Fortunately, Git doesn't allow adding anything _silently_, all changes are done via commits. If you really want to be notified of **every** commit and comment that happened in the kaitai-io org, use the GitHub API for public organization events: https://api.github.com/orgs/kaitai-io/events

I don't really understand why GitHub doesn't allow accessing the activity feed of an arbitrary org of which you're not a member directly in the web GUI or RSS format, but looking at the API response I link above, there is really all activity I see in my kaitai-io dashboard :-)

So if you're _determined_ to never miss a thing anymore, even the small and insignificant ones like `#pragma once`, give it a few hours and I believe that you can handcraft a Python script that generates an RSS feed or a fancy nice HTML+CSS human-friendly format or whatever output you'd like to. There are also numerous libraries to simplify the work with the GitHub API like https://github.com/PyGithub/PyGithub.

Just note that GitHub API has a rate limit of 60 requests per hour for unauthenticated and 5000 per hour for authenticated requests, which is enough though for almost any use. See https://developer.github.com/v3/#rate-limiting.

---

And my decision on the https://github.com/kaitai-io/kaitai_struct_compiler/commit/25fb1eee61d07bdc8b199445776836ce9a6606ef commit ref doesn't change, there is no place for it in the release highlights.